js/src/langchain.ts

```diff
@@ -0,0 +1,122 @@
+import { CallbackManager } from "@langchain/core/callbacks/manager";
+import { LangChainTracer } from "@langchain/core/tracers/tracer_langchain";
+import { Runnable, RunnableConfig } from "@langchain/core/runnables";
+
+import { RunTree } from "./run_trees.js";
+import { Run } from "./schemas.js";
+import {
+  TraceableFunction,
+  getCurrentRunTree,
+  isTraceableFunction,
+} from "./traceable.js";
+
+/**
+ * Converts the current run tree active within a traceable-wrapped function
+ * into a LangChain compatible callback manager. This is useful to handoff tracing
+ * from LangSmith to LangChain Runnables and LLMs.
+ *
+ * @param {RunTree | undefined} currentRunTree Current RunTree from within a traceable-wrapped function. If not provided, the current run tree will be inferred from AsyncLocalStorage.
+ * @returns {CallbackManager | undefined} Callback manager used by LangChain Runnable objects.
+ */
+export async function getLangchainCallbacks(
+  currentRunTree?: RunTree | undefined
+) {
+  const runTree: RunTree | undefined = currentRunTree ?? getCurrentRunTree();
+  if (!runTree) return undefined;
+
+  // TODO: CallbackManager.configure() is only async due to LangChainTracer
+  // factory being unnecessarily async.
+  let callbacks = await CallbackManager.configure();
+  if (!callbacks && runTree.tracingEnabled) {
+    callbacks = new CallbackManager();
+  }
+
+  let langChainTracer = callbacks?.handlers.find(
+    (handler): handler is LangChainTracer =>
+      handler?.name === "langchain_tracer"
+  );
+
+  if (!langChainTracer && runTree.tracingEnabled) {
+    langChainTracer = new LangChainTracer();
+    callbacks?.addHandler(langChainTracer);
+  }
+
+  const runMap = new Map<string, Run>();
+
+  // find upward root run
+  let rootRun = runTree;
+  const rootVisited = new Set<string>();
+  while (rootRun.parent_run) {
+    if (rootVisited.has(rootRun.id)) break;
+    rootVisited.add(rootRun.id);
+    rootRun = rootRun.parent_run;
+  }
+
+  const queue = [rootRun];
+  const visited = new Set<string>();
+
+  while (queue.length > 0) {
+    const current = queue.shift();
+    if (!current || visited.has(current.id)) continue;
+    visited.add(current.id);
+
+    runMap.set(current.id, current);
+    if (current.child_runs) {
+      queue.push(...current.child_runs);
+    }
+  }
+
+  if (callbacks != null) {
+    Object.assign(callbacks, { _parentRunId: runTree.id });
```

Comment:
This is to get around the fact that it's a private property?

Can we add a flag to add a setter method?

Reply:
In the https://github.com/langchain-ai/langchainjs/pull/5339 PR we already made this writable (while still keeping it private), this is needed for backwards compatibility